proposals/XXXX-compact.md

```diff
@@ -0,0 +1,280 @@
+---
+layout: default
+title: The Compact Programming Language
+parent: Hyperledger Improvement Proposals
+---
+
+# HIP Identifier
+<mark>**HIP identifier** a short description plus a serial number with a
+version.
+</mark>
+
+# Sponsor(s)
+- Parisa Ataei (Shielded) [<parisa.ataei@shielded.io>](mailto:parisa.ataei@shielded.io)
+- Bob Blessing-Hartley (Shielded) [<bob.blessing-hartley@shielded.io>](mailto:bob.blessing-hartley@shielded.io)
+- Joseph Denman (Shielded) [<joseph.denman@shielded.io>](mailto:joseph.denman@shielded.io)
+- Kent Dybvig (Shielded) [<kent.dybvig@shielded.io>](mailto:kent.dybvig@shielded.io)
+- Thomas Kerber (Shielded) [<thomas.kerber@shielded.io>](mailto:thomas.kerber@shielded.io)
+- Kevin Millikin (Shielded) [<kevin.millikin@shielded.io>](mailto:kevin.millikin@shielded.io)
+- Lucas Rosa (Shielded) [<lucas.rosa@shielded.io>](mailto:lucas.rosa@shielded.io)
+
+# Abstract
+**Compact** is a restricted programming language for privacy-preserving smart contracts in a decentralized blockchain.
+It was designed for the Midnight Network but it is intended to be suitable for any privacy-preserving blockchain with certain similar characteristics.
+
+# Context
+The Compact programming language provides a concrete implementation of the [Kachina programming model](https://iohk.io/en/research/library/papers/kachina-foundations-of-private-smart-contracts/) (Kerber, Kiayias, and Kohlweiss, 2021).
+Compact is used to implement privacy-preserving smart contracts with these characteristics:
+
+- contracts are compiled to **off-chain code** that implements the full contract with access to private data;
+- contracts are also compiled to **on-chain code** that implements (only) the contract's public state updates without access to private data;
+- transactions contain a **zero-knowledge (ZK) proof** of the off chain execution; and
+- the on-chain code is executed on chain if the blockchain can verify the ZK proof.
+
+The current implementation of Compact specifically targets the Midnight Network.
+The characteristics listed above are concretely realized in the following ways:
+
+- Compact contracts are compiled to off-chain JavaScript code;
+- on-chain code is implemented by the Midnight Network's [Impact Virtual Machine (VM)](https://docs.midnight.network/develop/how-midnight-works/impact);
+- an implementation of the BLS12-381 proving system is used for ZK proofs; and
+- the Midnight Network's blockchain is used to verify ZK proofs and execute the public part of the contract on chain.
+
+However, this is merely the specific implementation of Compact in the Midnight Network.
+The language itself is designed to be generically used in any privacy-preserving smart contract system that is sufficiently similar,
```

Comment:
Could you provide characteristics that would define what a "sufficiently similar" smart contract system would be?

Reply:
Yes, it's the characteristics described just above (off-chain code has access to private data, on-chain code does not, a ZK proof is constructed off chain and verified on chain).

I'll try to make that clear.